internals/feature_helpers.py

```diff
@@ -312,63 +312,149 @@ def get_in_milestone(milestone: int,
+        ndb.OR(Stage.stage_type == STAGE_BLINK_ORIGIN_TRIAL,
+            Stage.stage_type == STAGE_FAST_ORIGIN_TRIAL,
+            Stage.stage_type == STAGE_DEP_DEPRECATION_TRIAL))
+    desktop_origin_trial_future = q.fetch_async()
+
+    # Origin trial stages (Android) in this milestone.
+    q = Stage.query(Stage.milestones.android_first == milestone,
+        Stage.milestones.desktop_first == None,
+        ndb.OR(Stage.stage_type == STAGE_BLINK_ORIGIN_TRIAL,
+            Stage.stage_type == STAGE_FAST_ORIGIN_TRIAL,
+            Stage.stage_type == STAGE_DEP_DEPRECATION_TRIAL))
+    android_origin_trial_future = q.fetch_async()
+
+    # Origin trial stages (Webview) in this milestone.
+    q = Stage.query(Stage.milestones.webview_first == milestone,
+        Stage.milestones.desktop_first == None,
+        ndb.OR(Stage.stage_type == STAGE_BLINK_ORIGIN_TRIAL,
+            Stage.stage_type == STAGE_FAST_ORIGIN_TRIAL,
+            Stage.stage_type == STAGE_DEP_DEPRECATION_TRIAL))
+    webview_origin_trial_future = q.fetch_async()
+
+    # Dev trial stages (Desktop) in this milestone.
+    q = Stage.query(Stage.milestones.desktop_first == milestone,
+        ndb.OR(Stage.stage_type == STAGE_BLINK_DEV_TRIAL,
+            Stage.stage_type == STAGE_PSA_DEV_TRIAL,
+            Stage.stage_type == STAGE_FAST_DEV_TRIAL,
+            Stage.stage_type == STAGE_DEP_DEV_TRIAL))
+    desktop_dev_trial_future = q.fetch_async()
+
+    # Dev trial stages (Android) in this milestone.
+    q = Stage.query(Stage.milestones.android_first == milestone,
+        Stage.milestones.desktop_first == None,
+        ndb.OR(Stage.stage_type == STAGE_BLINK_DEV_TRIAL,
+            Stage.stage_type == STAGE_PSA_DEV_TRIAL,
+            Stage.stage_type == STAGE_FAST_DEV_TRIAL,
+            Stage.stage_type == STAGE_DEP_DEV_TRIAL))
+    android_dev_trial_future = q.fetch_async(None)
+
+    # Wait for all futures to complete and collect unique feature IDs.
+    desktop_shipping_ids = list(set(
+        [s.feature_id for s in desktop_shipping_future.result()]))
+    android_only_shipping_ids = list(set(
+        [s.feature_id for s in android_only_shipping_future.result()]))
+    desktop_origin_trials_ids = list(set(
+        [s.feature_id for s in desktop_origin_trial_future.result()]))
+    android_origin_trials_ids = list(set(
+        [s.feature_id for s in android_origin_trial_future.result()]))
+    webview_origin_trials_ids = list(set(
+        [s.feature_id for s in webview_origin_trial_future.result()]))
+    desktop_dev_trials_ids = list(set(
+        [s.feature_id for s in desktop_dev_trial_future.result()]))
+    android_dev_trials_ids = list(set(
+        [s.feature_id for s in android_dev_trial_future.result()]))
+
+    # Query for FeatureEntry entities that match the stage feature IDs.
+    # Querying with an empty list will raise an error, so check if each
+    # list is not empty first.
+    if desktop_shipping_ids:
```

Comment:
You could save some lines by introducing a tiny function that returns None or a future.  It would be called like:
desktop_shipping_future = _get_entries_by_id_async(desktop_shipping_ids)

Reply:
I went with this approach, and I agree that it's much more readable and less repetitive. 🙂